internal/pkg/push/push.go

```diff
@@ -0,0 +1,130 @@
+// ------------------------------------------------------------------------
+// SPDX-FileCopyrightText: Copyright © 2024 bomctl a Series of LF Projects, LLC
+// SPDX-FileName: internal/pkg/push/push.go
+// SPDX-FileType: SOURCE
+// SPDX-License-Identifier: Apache-2.0
+// ------------------------------------------------------------------------
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ------------------------------------------------------------------------
+package push
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/jdx/go-netrc"
+	"github.com/protobom/protobom/pkg/formats"
+	"github.com/protobom/protobom/pkg/sbom"
+
+	"github.com/bomctl/bomctl/internal/pkg/db"
+	"github.com/bomctl/bomctl/internal/pkg/options"
+	"github.com/bomctl/bomctl/internal/pkg/url"
+)
+
+var errUnsupportedURL = errors.New("failed to parse URL; see `bomctl push --help` for valid URL patterns")
+
+type (
+	Pusher interface {
+		url.Parser
+		Push(*sbom.Document, *url.ParsedURL, *url.BasicAuth) error
+		Name() string
+	}
+
+	Options struct {
+		*options.Options
+		Format   formats.Format
+		UseNetRC bool
+		UseTree  bool
+	}
+)
+
+func Push(sbomID, destPath string, opts *Options) error {
+	backend := db.NewBackend().
+		Debug(opts.Debug).
+		WithDatabaseFile(filepath.Join(opts.CacheDir, db.DatabaseFile)).
+		WithLogger(opts.Logger)
+
+	if err := backend.InitClient(); err != nil {
+		return fmt.Errorf("failed to initialize backend client: %w", err)
+	}
+
+	defer backend.CloseClient()
+
+	// Insert pushed document data into database.
+	document, err := backend.GetDocumentByID(sbomID)
+	if err != nil {
+		return fmt.Errorf("%w", err)
+	}
+
+	// Push externally referenced BOMs
+	return pushDocument(document, destPath, opts)
+}
+
+func NewPusher(sbomURL string) (Pusher, error) {
```

Comment:
Probably should have done this with fetcher as well, so it's called from outside this package as `pusher.New()` instead of `pusher.NewPusher()`

```suggestion
func New(sbomURL string) (Pusher, error) {
```